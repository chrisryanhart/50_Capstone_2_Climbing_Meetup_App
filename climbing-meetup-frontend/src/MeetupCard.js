import React from 'react';
import {Link} from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { Avatar } from '@material-ui/core';

const useStyles = makeStyles({
  root: {
    minWidth: 275,
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
});
// style={{ marginLeft: '5px'}}
export default function MeetupCard() {
  const classes = useStyles();
  const bull = <span className={classes.bullet}>â€¢</span>;

  return (
    <Card className={classes.root} style={{ marginTop: '10px'}}>
      <CardContent>
        <Typography variant="h5" component="h2" >
          Climbing at LEF 
        </Typography>
        <Typography variant="h7" component="h7" style={{ marginLeft: '5px'}}>
          <i>Date: 1/5/23</i>
          <br/>
        </Typography>
        <Typography variant="h7" component="h7" style={{ marginLeft: '5px'}}>
          <i>Time: 5:00 PM</i>
          <br/>
        </Typography>
        <Typography variant="h7" component="h7" style={{ marginLeft: '5px'}}>
          <i>Type: Lead Climbing</i>
          <br/>
        </Typography>
        <Typography variant="h7" component="h7" style={{ marginLeft: '5px'}}>
          <i>
            Description: Climb in the evening after work.  Bring your kids.
          </i>
        </Typography>
        <Typography variant="body2" component="div">
          <b>Organized by:</b> Spider Monkey
          <Avatar 
            alt="Spider Monkey" 
            src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
            />
        </Typography>
        <Typography variant="body2" sx={{display: 'inline-block'}}>
          <b>Attendees: </b> Chris
          <Avatar 
            alt="Spider Monkey" 
            src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
            />
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small">Join Meetup</Button>
        <Button size="small"><Link to="/meetups/1/manage">Manage</Link></Button>
      </CardActions>
    </Card>
  );
}
