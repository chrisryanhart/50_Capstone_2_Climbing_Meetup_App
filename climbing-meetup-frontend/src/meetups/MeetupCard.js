import React, { useContext, useEffect, useState } from 'react';
import {Link} from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { Avatar } from '@material-ui/core';
import CountContext from '../UserContext';
import ClimbMeetupApi from '../api';

const useStyles = makeStyles({
  root: {
    minWidth: 275,
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
});

export default function MeetupCard({details}) {
  const classes = useStyles();

  const {userMeetups, currUserId} = useContext(CountContext);

  const currUserIdInt = Number(currUserId);

  // check is the current user created the meetup
  const isCreator = currUserIdInt === details.creator_user_id ? true : false;

  // const attendeeList = details.attendees;
  const [attendeeList, setAttendeeList] = useState(details.attendees);
  // const [attendees, setAttendees] = useState([]);
  // const attendeeStatus = {status: ''};
  // const INITIAL_ATTENDEE_STATUS = {status: ''};
  // const [attendeeStatus, setAttendeeStatus] = useState(INITIAL_ATTENDEE_STATUS);

  const attendeeStatus = {status: ''};

  const attendees = attendeeList.map(attendee =>{

    if(!isCreator){
      if(currUserIdInt === attendee.id){
        attendeeStatus.status = attendee.status;
      }
    }

    if(attendee.status==='approved'){
      return (
      <div >
        <Link to={`/users/${attendee.id}`}>
            <div style={{display: 'flex'}}>
              <Avatar 
                      alt="Spider Monkey" 
                      src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
                      />
              <Typography data-testid="attendee" variant="body2" style={{marginLeft: '10px',marginTop: '10px',alignItems: 'center'}}>
                  {attendee.name}
                  <br/>
              </Typography>
            </div>
        </Link>

      </div>);
      }
});


  // set initial meetupJoinStatus after checking is user is in the attendee list
  const [meetupJoinStatus, setMeetupJoinStatus] = useState(attendeeStatus.status);
  // const [pageLoaded, setPageLoaded] = useState(false);

    const handleJoin = async () => {
      let res = await ClimbMeetupApi.joinMeetup(details.id);
      setMeetupJoinStatus(res.join_request_status);
    }

    const leaveMeetup = async () => {
      // remove attendee from state
      // call database
      let res = await ClimbMeetupApi.leaveMeetup(details.id);
      console.log('test');
      // setAttendee list
      setMeetupJoinStatus('');
      setAttendeeList(attendees => attendees.filter(attendee => attendee.id !== currUserId));
    }

  // handle which button to display based on attendee status
  let joinStatusButton;

  if(meetupJoinStatus===''){
    joinStatusButton = <Button onClick={handleJoin} size="small">Join Meetup</Button>;
  }
  if(meetupJoinStatus==='pending'){
    joinStatusButton = <Button size="small">Pending</Button>;
  }else if(meetupJoinStatus==='approved'){
    joinStatusButton = <Button onClick={leaveMeetup} size="small">Leave Meetup</Button>;
  }

  // modify date format
  // can get the utc_date_time from the database

  const rawDateFormat = new Date(details.date);
  const dateStr = rawDateFormat.toLocaleDateString('en-US');

  // use the following for utc_date_time conversions
  // const rawDate = new Date(details.utc_date_time);
  // const date = rawDate.toLocaleDateString('en-US');
  // const time = rawDate.toLocaleTimeString('en-US')

  return (
    <Card className={classes.root} style={{ marginTop: '10px'}}>
      <CardContent>
        <Typography variant="h5" component="h2" >
          Climb at {details.location_name} 
        </Typography>
        <Typography variant="h6" component="h6" style={{ marginLeft: '5px'}}>
          <i>Date: {dateStr}</i>
          <br/>
        </Typography>
        <Typography variant="h6" component="h6" style={{ marginLeft: '5px'}}>
          <i>Time: {details.time}</i>
          <br/>
        </Typography>
        <Typography variant="h6" component="h6" style={{ marginLeft: '5px'}}>
          <i>Type: Lead Climbing</i>
          <br/>
        </Typography>
        <Typography variant="h6" component="h6" style={{ marginLeft: '5px'}}>
          <i>
            Description: {details.description}
          </i>
        </Typography>
        <Typography className={classes.title} gutterBottom>
          <b>Organized by: </b> 
        </Typography>
        <div>
            <Link to={`/users/${details.creator_user_id}`}>
              <div style={{display: 'flex'}}>
                  <Avatar 
                          alt="Spider Monkey" 
                          src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
                          />
                  <Typography data-testid="creator-name" variant="body2" style={{marginLeft: '10px',marginTop: '10px',alignItems: 'center'}}>
                    {details.creator_name}
                  </Typography>
              </div>
            </Link>
        </div>
        <Typography className={classes.title} gutterBottom>
          <b>Attendees: </b>
        </Typography>
        {attendees}
      </CardContent>
      <CardActions>
        {!isCreator && joinStatusButton}
        {isCreator && <Button size="small"><Link to={`/meetups/${details.id}/manage`}>Manage</Link></Button>}
      </CardActions>

    </Card>
  );
}
