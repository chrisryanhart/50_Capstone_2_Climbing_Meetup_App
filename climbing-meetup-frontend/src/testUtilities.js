import React from "react";
import UserContext from "./UserContext";

// const demoUser = {
//   username: "testuser",
//   first_name: "testfirst",
//   last_name: "testlast",
//   email: "test@test.net",
//   photo_url: null,
// };

// token
const userToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJ0ZXN0dXNlcjEiLCJpYXQiOjE2NzQ0MzYyMDZ9.m7tigic9I4D58MYFqDDBmc7cnZTSB5qtLkij5-Ol36A';

const regMock = jest.fn();

const UserProvider =
    ({ children, currUserId = 1, token=userToken, registerUser={regMock}}) => (
    <UserContext.Provider value={{ currUserId, token, registerUser }}>
      {children}
    </UserContext.Provider>
);

export { UserProvider,regMock };