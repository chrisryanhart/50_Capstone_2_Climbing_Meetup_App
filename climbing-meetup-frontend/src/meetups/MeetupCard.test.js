import React from "react";
import { render } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import { UserProvider, regMock } from "../testUtilities";
import { MemoryRouter } from "react-router-dom";
import MeetupCard from "./MeetupCard";

it("matches snapshot", function () {

    const meetupDetails1  = {
        creator_name: 'spider man',
        creator_user_id: 1,
        date: '2023-01-05T05:00:00.000Z',
        description: 'Climb and have some beers',
        duration: 3,
        id: 1,
        location_name: 'Red River Gorge',
        time: '10:00:00',
        attendees:[{id: 2, name: 'wonder woman', status: 'pending'},{id: 3, name: 'batman', status: 'approved'},{id: 19, name: 'spongebob', status: 'pending'}]
    }

    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <MeetupCard details={meetupDetails1} />
            </UserProvider>
        </MemoryRouter>

        );
    expect(asFragment()).toMatchSnapshot();

  });

it("renders without crashing", function() {

    const meetupDetails1  = {
        creator_name: 'spider man',
        creator_user_id: 1,
        date: '2023-01-05T05:00:00.000Z',
        description: 'Climb and have some beers',
        duration: 3,
        id: 1,
        location_name: 'Red River Gorge',
        time: '10:00:00',
        attendees:[{id: 2, name: 'wonder woman', status: 'pending'},{id: 3, name: 'batman', status: 'approved'},{id: 19, name: 'spongebob', status: 'pending'}]
    }

    render(
        <MemoryRouter>
            <UserProvider>
                <MeetupCard details={meetupDetails1} />
            </UserProvider>
        </MemoryRouter>

        );
});


it("loads current user meetup details", function() {

    const meetupDetails2 = {
        attendees: [{id: 1, name: 'spider man', status: 'approved'}],
        creator_name: 'wonder woman',
        creator_user_id: 2,
        date:'2023-01-10T05:00:00.000Z',
        description: 'Meetup after work',
        duration: 4,
        id: 2,
        location_name: 'Climb Lawrence',
        time: '17:00:00'
    }
    const { getByTestId, getByText } = render(
      <MemoryRouter>
        <UserProvider >
            <MeetupCard details={meetupDetails2} />
        </UserProvider>
      </MemoryRouter>
      );

    expect(getByTestId('attendee').textContent).toEqual('spider man');
    expect(getByTestId('creator-name').textContent).toEqual('wonder woman'); 
    expect(getByText('Leave Meetup')).toBeInTheDocument();

});


it("loads current user meetup details", function() {
    const meetupDetails1  = {
        creator_name: 'spider man',
        creator_user_id: 1,
        date: '2023-01-05T05:00:00.000Z',
        description: 'Climb and have some beers',
        duration: 3,
        id: 1,
        location_name: 'Red River Gorge',
        time: '10:00:00',
        attendees:[{id: 2, name: 'wonder woman', status: 'pending'},{id: 3, name: 'batman', status: 'approved'},{id: 19, name: 'spongebob', status: 'pending'}]
    }

    const meetupDetails2 = {
        attendees: [{id: 1, name: 'spider man', status: 'approved'}],
        creator_name: 'wonder woman',
        creator_user_id: 2,
        date:'2023-01-10T05:00:00.000Z',
        description: 'Meetup after work',
        duration: 4,
        id: 2,
        location_name: 'Climb Lawrence',
        time: '17:00:00'
    }
    const { getByTestId, getByText } = render(
      <MemoryRouter>
        <UserProvider >
            <MeetupCard details={meetupDetails1} />
        </UserProvider>
      </MemoryRouter>
      );

    expect(getByTestId('attendee').textContent).toEqual('batman');
    expect(getByTestId('creator-name').textContent).toEqual('spider man'); 
    expect(getByText('Manage')).toBeInTheDocument();

});
