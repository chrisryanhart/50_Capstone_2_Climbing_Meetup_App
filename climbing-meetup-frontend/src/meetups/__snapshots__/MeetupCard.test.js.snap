// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
    style="margin-top: 10px;"
  >
    <div
      class="MuiCardContent-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5"
      >
        Climb at Red River Gorge
      </h2>
      <h6
        class="MuiTypography-root MuiTypography-h6"
        style="margin-left: 5px;"
      >
        <i>
          Date: 1/5/2023
        </i>
        <br />
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-h6"
        style="margin-left: 5px;"
      >
        <i>
          Time: 10:00:00
        </i>
        <br />
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-h6"
        style="margin-left: 5px;"
      >
        <i>
          Type: Lead Climbing
        </i>
        <br />
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-h6"
        style="margin-left: 5px;"
      >
        <i>
          Description: Climb and have some beers
        </i>
      </h6>
      <p
        class="MuiTypography-root makeStyles-title-3 MuiTypography-body1 MuiTypography-gutterBottom"
      >
        <b>
          Organized by: 
        </b>
      </p>
      <div>
        <a
          href="/users/1"
        >
          <div
            style="display: flex;"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular"
            >
              <img
                alt="Spider Monkey"
                class="MuiAvatar-img"
                src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2"
              data-testid="creator-name"
              style="margin-left: 10px; margin-top: 10px; align-items: center;"
            >
              spider man
            </p>
          </div>
        </a>
      </div>
      <p
        class="MuiTypography-root makeStyles-title-3 MuiTypography-body1 MuiTypography-gutterBottom"
      >
        <b>
          Attendees: 
        </b>
      </p>
      <div>
        <a
          href="/users/3"
        >
          <div
            style="display: flex;"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular"
            >
              <img
                alt="Spider Monkey"
                class="MuiAvatar-img"
                src="https://firebasestorage.googleapis.com/v0/b/climbing-meetup-app.appspot.com/o/sean-benesh-VnmbcgAfL3Q-unsplash.jpg?alt=media&token=9f5685b0-3529-40c3-98d2-5b54b1b09825"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2"
              data-testid="attendee"
              style="margin-left: 10px; margin-top: 10px; align-items: center;"
            >
              batman
              <br />
            </p>
          </div>
        </a>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <a
            href="/meetups/1/manage"
          >
            Manage
          </a>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
